primitives/runtime/src/offchain/pollable.rs

```diff
@@ -0,0 +1,439 @@
+			Poll::Pending => return Poll::Pending,
+			Poll::Ready(()) => {},
+		}
+
+		let now = sp_io::offchain::timestamp();
+		let status = sp_io::offchain::http_response_wait(&[id], Some(now));
+
+		let code = match status.as_slice() {
+			[HttpRequestStatus::Finished(code)] => *code,
+			[HttpRequestStatus::IoError] => return Poll::Ready(Err(HttpError::IoError)),
+			[HttpRequestStatus::Invalid] => return Poll::Ready(Err(HttpError::Invalid)),
+			[HttpRequestStatus::DeadlineReached] => {
+				// Internal logic error: We were signaled as ready but
+				// the request somehow internally timed out.
+				// Return an error instead of panicking
+				return Poll::Ready(Err(HttpError::IoError));
+			},
+			[] | [_, _, ..] => unreachable!("waiting for a single ID should give exactly one; qed"),
+		};
+
+		let headers = sp_io::offchain::http_response_headers(id);
+
+		Poll::Ready(Ok(HttpResponse {
+			code,
+			headers,
+			body: HttpBodyFuture {
+				id,
+				buf: Some(vec![0; 1024]),
+				len: 0,
+			},
+		}))
+	}
+}
+
+
+/// A future type that resolves to an HTTP response body.
+///
+/// Part of the [`HttpResponse`] value.
+#[derive(Debug)]
+pub struct HttpBodyFuture {
+	id: HttpRequestId,
+	buf: Option<Vec<u8>>,
+	len: usize,
+}
+
+impl Future for HttpBodyFuture {
+	type Output = Box<[u8]>;
+
+	fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+		// Help the borrow checker see disjoint field mutable borrow through Pin
+		let this = &mut *self;
+		let (id, len) = (this.id, &mut this.len);
+		let self_buf = this.buf.as_mut().expect("HttpBodyFuture polled after value taken");
+
+		loop {
+			let buf = &mut self_buf[*len..];
+			// Ensure that we can always write something, so that reading 0 bytes
+			// means EOF rather than inability to write more
+			debug_assert_ne!(buf.len(), 0);
+
+			// Read body in a non-blocking fashion with present timestamp as deadline
+			let now = sp_io::offchain::timestamp();
+			match sp_io::offchain::http_response_read_body(id, buf, Some(now)) {
+				Ok(0) => {
+					let mut buf = this.buf.take()
+						.expect("HttpBodyFuture polled after value taken");
+					buf.truncate(*len);
+
+					return Poll::Ready(buf.into_boxed_slice());
+				},
+				Ok(bytes_read) => {
+					// Grow the temporary buffer if it's full
+					if bytes_read as usize == buf.len() {
+						let growth = self_buf.len() / 2 * 3; // 1.5
+						let new_len = self_buf.len() + growth;
+						self_buf.resize_with(new_len, Default::default);
+					}
+					*len += bytes_read as usize;
+
+					continue;
+				},
+				Err(HttpError::DeadlineReached) => {
+					let fut = HostFuture::from(id);
+					pin_mut!(fut);
+					if let Poll::Ready(()) = fut.poll(cx) {
+						// More work is ready by now, reschedule to poll again
+						cx.waker().wake_by_ref();
+					}
+
+					return Poll::Pending;
+				},
+				Err(HttpError::IoError) | Err(HttpError::Invalid) => panic!(
+					"HttpBodyFuture is created for HTTP requests that are Finished \
```

Comment:
Are you sure we don't return `IoError` in case the connection is dropped or sth?

Reply:
Good catch! Updated